### ISSUE TITLE
prefer-array-literal: Error when using Array in a type annotation

### ISSUE THREAD

**Issue by nfriend** [Microsoft/tslint-microsoft-contrib#110]
Perhaps this is the intended behavior, but from the description of the rule, the `prefer-array-literal` rule looks like its purpose is to disallow using `Array` when instantiating new arrays:

```typescript
new Array();
new Array(4,5);
// etc...
```

However, this rule also disallows using `Array` as a type parameter, which is one of the recommended ways of declaring Array types in [the handbook](http://www.typescriptlang.org/docs/handbook/basic-types.html#array):

```typescript
// these throw "Replace generic-typed Array with array literal" linting error

let myArray: Array<MyObject> = [];

interface MyInterface {
    myArray: Array<number>;
}
```

I expected this rule to throw errors on the first set of examples above but allow the second set.

**Comment by HamletDRC** [comment]
Good find! Our coding standard says to prefer the literal syntax, even when declaring the array type. 

We can add a parameter to the rule that will let you turn this on. 

Why do you prefer 

    let myArray: Array<MyObject> = []

instead of

    let myArray: MyObject[] = [];

The 2nd is shorter and gives your code base one single, consistent way to declare array types. If it were me, I would just follow the rule's advice. But that's a bit unfair because I wrote the rule :)

**Comment by nfriend** [comment]
Thanks for the response, @HamletDRC.  I don't necessarily prefer the `Array<>` syntax, but my team's existing codebase uses it heavily.  Adding an option to ignore this error when the syntax is part of a type declaration would make it much easier to integrate this rule into our project.

**Comment by webberwang** [comment]
I prefer the first method because it reads more fluently. Array of Objects as opposed to Object Array

**Comment by jburtondev** [comment]
Try this: `(Type1 | Type2)[]`

**Comment by ogiexela** [comment]
how can one specify the size of the array?

const foo = new Array(10)
?

**Comment by sangheestyle** [comment]
@ogiexela You can use `const foo = Array(10)` instead of `const foo = new Array(10)`. The result will be the same but `prefer-array-literal` will be disappeared immediately.

https://stackoverflow.com/a/8205709/568901

**Comment by JoshuaKGoldberg** [comment]
Oh that sounds like a bug in the rule...

**Comment by sangheestyle** [comment]
@JoshuaKGoldberg OK. how can I help this issue?

**Comment by JoshuaKGoldberg** [comment]
That'd be great! Would you mind filing a new issue describing what you're seeing regarding being allowed to use `= Array(10)` but not `new Array(10)`? Including the normal information included with issues (version, code snippets, etc).

**Comment by sangheestyle** [comment]
@JoshuaKGoldberg Yes, I will do. I just checked which ts-lint actually does `prefer-array-literal` and I found it is by `tslint-microsoft-contrib`.

https://github.com/palantir/tslint/issues/2946#issuecomment-310581618

I am going to test it with up-to-date version of `tslint-microsoft-contrib` then I will create issue.

**Comment by sangheestyle** [comment]
@jonascarlsonschmidt Created. Thanks for guiding me.

https://github.com/Microsoft/tslint-microsoft-contrib/issues/839
